L1Trigger/TrackerTFP/src/KalmanFilter.cc

```diff
@@ -0,0 +1,398 @@
+          if (track->hitPattern(nextLayer)) {
+            stubNext = track->layerStub(nextLayer);
+            break;
+          }
+        }
+      }
+    }
+    if (stubNext) {
+      states_.emplace_back(state, stubNext);
+      state = &states_.back();
+    } else
+      state = nullptr;
+  }
+
+  // best state selection
+  void KalmanFilter::accumulator(deque<State*>& stream) {
+    // accumulator delivers contigious stream of best state per track
+    // remove gaps and not final states
+    stream.erase(remove_if(stream.begin(), stream.end(), [this](State* state){ return !state || state->hitPattern().count() < setup_->kfMaxLayers(); }), stream.end());
+    /*auto toLessPS = [this](State* state) {
+      int nPS(0);
+      for (const StubKF& stub : state->stubs())
+        if (setup_->psModule(stub.ttStubRef()))
+          nPS++;
+      return nPS < 2;
+    };
+    stream.erase(remove_if(stream.begin(), stream.end(), toLessPS), stream.end());*/
+    // update chi2
+    for (State* state : stream)
+      state->finish();
+    // sort in number of skipped layers
+    auto lessSkippedLayers = [](State* lhs, State* rhs) { return lhs->numSkippedLayers() < rhs->numSkippedLayers(); };
+    stable_sort(stream.begin(), stream.end(), lessSkippedLayers);
+    // sort in number of consistent stubs
+    auto moreConsistentLayers = [](State* lhs, State* rhs) { return lhs->numConsistentLayers() > rhs->numConsistentLayers(); };
+    stable_sort(stream.begin(), stream.end(), moreConsistentLayers);
+    // sort in track id
+    stable_sort(stream.begin(), stream.end(), [](State* lhs, State* rhs){ return lhs->trackId() < rhs->trackId(); });
+    // keep first state (best due to previous sorts) per track id
+    stream.erase(unique(stream.begin(), stream.end(), [](State* lhs, State* rhs){ return lhs->track() == rhs->track(); }), stream.end());
+    //for (State* state : stream)
+      //cout << state->x0() << " " << state->x1() << " " << dataFormats_->chosenRofPhi() << endl;
+  }
+
+  // updates state
+  void KalmanFilter::update(State*& state) {
+    // R-Z plane
+    const double H12 = H12_->digi(state->H12());
+    const double m1 = m1_->digi(state->m1());
+    const double v1 = v1_->digi(state->v1());
+    double x2 = x2_->digi(state->x2());
+    double x3 = x3_->digi(state->x3());
+    double C22 = C22_->digi(state->C22());
+    double C23 = C23_->digi(state->C23());
+    double C33 = C33_->digi(state->C33());
+    v1_->updateRangeActual(v1);
+    C22_->updateRangeActual(C22);
+    C23_->updateRangeActual(C23);
+    C33_->updateRangeActual(C33);
+    H12_->updateRangeActual(H12);
+    m1_->updateRangeActual(m1);
+    const double r1C = x3_->digi(m1  - x3);
+    const double r1 = r1_->digi(r1C - x2 * H12);
+    const double S12 = S12_->digi(C23 + H12 * C22);
+    const double S13 = S13_->digi(C33 + H12 * C23);
+    const double R11C = S13_->digi(v1 + S13);
+    const double R11 = R11_->digi(R11C + H12 * S12);
```

Comment:
Add comment "Cov. matrix of predicted residuals R = V+H*C*Ht = C+H*St"

Reply:
done